index.js

```diff
@@ -1050,131 +1136,209 @@
     return function(x) { return contravariant (f (x)); };
   }
 
+  var staticImplementations = {
+    String: {
+      empty: String$empty
+    },
+    Array: {
+      empty: Array$empty,
+      of: Array$of,
+      chainRec: Array$chainRec,
+      zero: Array$zero
+    },
+    Object: {
+      empty: Object$empty,
+      zero: Object$zero
+    },
+    Function: {
+      id: Function$id,
+      of: Function$of,
+      chainRec: Function$chainRec
+    }
+  };
+
   function staticMethod(name, typeRep) {
-    switch (functionName (typeRep) + '.' + name) {
-      case 'String.fantasy-land/empty':
-        return String$empty;
-      case 'Array.fantasy-land/empty':
-        return Array$empty;
-      case 'Array.fantasy-land/of':
-        return Array$of;
-      case 'Array.fantasy-land/chainRec':
-        return Array$chainRec;
-      case 'Array.fantasy-land/zero':
-        return Array$zero;
-      case 'Object.fantasy-land/empty':
-        return Object$empty;
-      case 'Object.fantasy-land/zero':
-        return Object$zero;
-      case 'Function.fantasy-land/id':
-        return Function$id;
-      case 'Function.fantasy-land/of':
-        return Function$of;
-      case 'Function.fantasy-land/chainRec':
-        return Function$chainRec;
-      default:
-        return null;
+    switch (typeRep) {
+      case String: return staticImplementations.String[name];
+      case Array: return staticImplementations.Array[name];
+      case Object: return staticImplementations.Object[name];
+      case Function: return staticImplementations.Function[name];
+      case null: /* fall through */
+      case undefined: return undefined;
+    }
+
+    var prefixedName = 'fantasy-land/' + name;
+    if (typeof typeRep[prefixedName] === 'function') {
+      return typeRep[prefixedName];
+    }
+
+    if (typeof typeRep !== 'function') return;
+
+    switch (functionName (typeRep)) {
+      case 'String': return staticImplementations.String[name];
+      case 'Array': return staticImplementations.Array[name];
+      case 'Object': return staticImplementations.Object[name];
+      case 'Function': return staticImplementations.Function[name];
+    }
```

Comment:
This block is here for when the constructor passed was taken from a different global scope (or context) than the one available to this module. I'm not sure this is relevant at all for static methods, because the user passes the constructor directly, so they would have to do some weird stuff to obtain a constructor from one context, and pass it to Z in another.